module3/lessons/refactoring_api_consumption.md

```diff
@@ -0,0 +1,263 @@
+
+<section class="dropdown">
+### Possible Design Pattern #1: Gateways
+
+[Gateway classes](https://mattbrictson.com/blog/gateway-pattern) are very common across many frameworks. They are also referred to as a *service* but this is a very overloaded term in Rails and software development as a whole. This section will refer to these classes as *gateways* but you're welcome to name them however you'd prefer.
+
+Gateways typically encapsulate the logic used for interacting with an external API. Let's refactor our controller to use a gateway. 
+
+*app/controllers/images_controller.rb*
+```ruby
+class Api::V1::ImagesController < ApplicationController
+  def show
+    artist = params[:artist]
+
+    first_photo = ImageGateway.get_first_image(artist)
+    
+    formatted_json = {
+      id: nil,
+      type: "image",
+      attributes: {
+        image_url: first_photo[:url],
+        photographer: first_photo[:photographer],
+        photographer_url: first_photo[:photographer_url],
+        alt_text: first_photo[:alt]
+      }
+    }
+
+    render json: { data: formatted_json }
+  end
+end
+```
+
+Now our Gateway class might look like this:
+```ruby
+class ImageGateway
+  def self.get_first_image(query_term)
+    conn = Faraday.new(url: "https://api.pexels.com") do |faraday|
+      faraday.headers["Authorization"] = Rails.application.credentials.pexels[:key]
+    end
+
+    response = conn.get("/v1/search", { query: query_term })
+    # OR response = conn.get("/v1/search?query=#{artist})
+
+    json = JSON.parse(response.body, symbolize_names: true)
+    json[:photos][0]
+  end
+end
+```
+
+Notice how much logic that extracted out of the controller! Hooray! 
+
+If we are making calls to more than one Pexels endpoint in our application, we could also consider adding additional class methods to our Gateway and extacting the Faraday connection set up into a helper method that can be called in each method.
+
+```ruby
+class ImageGateway
+  def self.get_first_image(query_term)
+    response = connection.get("/v1/search", { query: query_term })
+
+    json = JSON.parse(response.body, symbolize_names: true)
+    json[:photos][0]
+  end
+
+  private
+  
+  def self.connection
+    Faraday.new(url: "https://api.pexels.com") do |faraday|
+      faraday.headers["Authorization"] = Rails.application.credentials.pexels[:key]
+    end
+  end
+end
+```
+
+
+We also know that serializers are a very helpful tool for managing the presentation of our data. Once we extract the JSON response formatting out of the controller, our controller action looks even nicer!
+
+*app/controllers/images_controller.rb*
+```ruby
+class Api::V1::ImagesController < ApplicationController
+  def show
+    artist = params[:artist]
+
+    first_photo = ImageGateway.get_first_image(artist)
+
+    render json: ImageSerializer.format_image_response(first_photo)
+  end
+end
+```
+
+This is a huge improvement! 
+
+</section>
+
+<section class="dropdown">
+### Possible Refactor #2: POROs
```

Comment:
Nit pick: The way these two options are worded, I was originally thinking I could refactor the original controller with #1 OR #2, but #2 is more of an addition to #1. Maybe consider wording this not as "Potential Refactor #1", but instead more in terms of steps? Like "Potential Refactor Step 1", then Step 2?

Reply:
ooh I'll change to "part 1" and "part 2"